core/commands/publish.go

```diff
@@ -205,6 +205,10 @@ func keylookup(n *core.IpfsNode, k string) (crypto.PrivKey, error) {
 		return nil, err
 	}
 
+	if k == n.Identity.Pretty() {
```

Comment:
I'd put this in https://github.com/ipfs/go-ipfs/blob/master/core/core.go#L689 (`n.GetKey`). This way it's also fixing `name.Publish` in coreapi - https://github.com/ipfs/go-ipfs/blob/master/core/coreapi/name.go#L142

Reply:
It's a good point that we should avoid duplicating this check in coreapi, but is it appropriate for `GetKey` to concern itself with checking for something that isn't a key name? The reason this fix is needed seems to be that the `self` key isn't returned from `n.Repo.Keystore().List()`. (Notice the code adds it manually in https://github.com/ipfs/go-ipfs/blob/master/core/commands/keystore.go#L192-L194)

Do you think it'd be better to fix it there instead?

Reply:
Right.. It isn't the best place (it's accepting the key _name_). Putting it in the return of `n.Repo.Keystore().List()` would imply that it is stored in keystore which it is not. The commands here are going to get rewritten to use coreapi eventually, so I think it is fine to duplicate this check to https://github.com/ipfs/go-ipfs/blob/master/core/coreapi/name.go#L142.

Reply:
Yeah, certainly we want to avoid duplicating this fix in coreapi, but should `GetKey` be concerned with something that isn't a key name?

I notice that every place that calls `Keystore.List()` either manually adds the self key (https://github.com/ipfs/go-ipfs/blob/master/core/commands/keystore.go#L184-L192, https://github.com/ipfs/go-ipfs/blob/master/core/coreapi/key.go#L97-L105), or currently has a bug because it doesn't manually add it. Would it be appropriate to change the `Keystore.List()` (and, I assume, `Keystore.Get()`) to return and recognize the self key?

Reply:
I'm not sure if this is the correct way, though it feels the cleanest to me

@Stebalien @Kubuxu Is there any reason the self key isn't handled by keystore methods?